File: src/main/java/com/leet/medium/MergeIntervals.go
package main

import "sort"

func merge(intervals [][]int) [][]int {
	sort.Slice(intervals, func(i, j int) bool {
		return intervals[i][0] < intervals[j][0]
	})

	begin, end := intervals[0][0], intervals[0][1]
	ans := [][]int{}
	for _, interval := range intervals {
		if end < interval[0] {
			ans = append(ans, []int{begin, end})
			begin, end = interval[0], interval[1]
		} else {
			if interval[1] > end {
				end = interval[1]
			}
		}
	}

	return append(ans, []int{begin, end})
}

File: src/main/java/com/leet/medium/MergeIntervals.ts
// https://leetcode.com/problems/merge-intervals/

function merge(intervals: number[][]): number[][] {
	intervals.sort((a, b) => a[0] - b[0]);
	return intervals.reduce((acc, cur) => {
		if (acc.length === 0) {
			acc.push(cur);
		} else {
			const prev = acc[acc.length - 1];
			if (prev[1] < cur[0]) {
				acc.push(cur);
			} else {
				if (prev[1] < cur[1]) {
					prev[1] = cur[1];
				}
			}
		}
		return acc;
	}, [] as number[][]);
};
